### PR TITLE
fix: oncePer condition only checked when `when` condition is true (#71)

### LINKED ISSUE #71: oncePer condition logs expected error
In argocd notifications, I use the following trigger to filter notifications with the `oncePer` option.

```yaml
    trigger.on-deployed: |
      - description: Application is synced and healthy. Triggered once per commit.
        when: app.status.health.status == 'Healthy' and app.status.operationState != nil and app.status.operationState.phase in ['Succeeded']
        oncePer: app.status.operationState.syncResult.revision
        send:
        - app-deployed
```

This returns the following error in the logs

```
{"level":"error","msg":"failed to execute oncePer condition: cannot fetch syncResult from \u003cnil\u003e (1:27)\n | app.status.operationState.syncResult.revision\n | ..........................^","time":"2022-01-28T15:41:46Z"}
```

### Solution
I think the `oncePer` condition should not be evaluated when the `when` condition evaluate to false.

https://github.com/argoproj/notifications-engine/blob/b9d7d6c233cc0c52b9c731f2ce1e909794c4347c/pkg/triggers/service.go#L94

### PR DESCRIPTION
**Pull request by lol3909**
This change makes it that the oncePer condition only gets check if the `when` condition is true.
fixes #71

### CODE REVIEW

**Comment by mubarak-j** [comment]
@pasha-codefresh would you be able to review this PR?

**Comment by lol3909** [comment]
Hi, any news on this? :) @mubarak-j or @pasha-codefresh

**Comment by mubarak-j** [comment]
I pinged the maintainers on Slack https://cloud-native.slack.com/archives/C01TSERG0KZ/p1696269136126209
I'm just an end-user that hoping this long-standing annoying bug gets resolved, pretty sure will help save in log costs 💸 😄

File: pkg/triggers/service.go [hunk 1]
```diff
@@ -81,22 +81,29 @@ func (svc *service) Run(triggerName string, vars map[string]interface{}) ([]Cond
...
 		if prog, ok := svc.compiledConditions[condition.When]; !ok {
-			return nil, fmt.Errorf("trigger configiration has changed after initialization")
+			return nil, fmt.Errorf("trigger configuration has changed after initialization")
 		} else if val, err := expr.Run(prog, vars); err == nil {
 			boolRes, ok := val.(bool)
 			conditionResult.Triggered = ok && boolRes
+			whenResult = conditionResult.Triggered
 		} else {
 			log.Errorf("failed to execute when condition: %+v", err)
+			whenResult = false
```
Comment (pasha-codefresh) [inline pkg/triggers/service.go:94]: whenResult can be false by default, so you will not need write this line

File: pkg/triggers/service.go [hunk 2]
```diff
@@ -81,22 +81,29 @@ func (svc *service) Run(triggerName string, vars map[string]interface{}) ([]Cond
 			Templates: condition.Send,
 			Key:       fmt.Sprintf("[%d].%s", i, hash(condition.When)),
 		}
-
+		var whenResult bool
+		log.Debug("This is the version with the whenResult variable")
```
Comment (pasha-codefresh) [inline pkg/triggers/service.go:85]: i think this log is redundant

**Review by pasha-codefresh — CHANGES_REQUESTED** [review]

**Comment by pasha-codefresh** [comment]
Hey @lol3909 , added few small comments

**Pull request merged**